polygeist/lib/polygeist/Passes/ConvertPolygeistToLLVM.cpp

```diff
@@ -90,84 +84,131 @@ struct SubIndexOpLowering : public ConvertOpToLLVMPattern<SubIndexOp> {
+
+    // When the converted source operand element type is a struct, the original
+    // memref element type represents a SYCL type. Handle the non-SYCL case
+    // first.
+    if (!convSourceElemType.isa<LLVM::LLVMStructType>()) {
+      assert(
+          convViewElemType ==
+              prev.getType().cast<LLVM::LLVMPointerType>().getElementType() &&
+          "Expecting the element types to match");
+      auto memRefDesc = createMemRefDescriptor(
+          loc, viewMemRefType, targetMemRef.allocatedPtr(rewriter, loc),
+          rewriter.create<LLVM::GEPOp>(loc, prev.getType(), prev, idxs), sizes,
+          strides, rewriter);
+
+      rewriter.replaceOp(subViewOp, {memRefDesc});
       return success();
     }
 
-    assert(getTypeConverter()->convertType(viewMemRefType.getElementType()) ==
-               prev.getType().cast<LLVM::LLVMPointerType>().getElementType() &&
-           "Expecting the element types to match");
-    MemRefDescriptor nexRef = createMemRefDescriptor(
-        loc, viewMemRefType, targetMemRef.allocatedPtr(rewriter, loc),
-        rewriter.create<LLVM::GEPOp>(loc, prev.getType(), prev, idxs), sizes,
-        strides, rewriter);
-
-    rewriter.replaceOp(subViewOp, {nexRef});
+    // SYCL case
+    assert(sourceMemRefType.getRank() == viewMemRefType.getRank() &&
+           "Expecting the input and output MemRef ranks to be the same");
+    // Note:in MLIRASTConsumer::getMLIRType(), a memref of struct type is
+    // generated only for a struct that has at least one member of SYCL type,
+    // otherwise a llvm pointer type is generated instead of a memref).
+    assert((sycl::isSYCLType(sourceElemType) ||
+            sourceElemType.isa<LLVM::LLVMStructType>() &&
+                any_of(sourceElemType.cast<LLVM::LLVMStructType>().getBody(),
+                       [](const Type &memType) {
+                         return sycl::isSYCLType(memType);
+                       })) &&
+           "the source memref element type should be either a SYCL type, or "
+           "a struct containing at least a member of SYCL type");
+
+    SmallVector<Value, 4> indices;
+    computeIndices(convSourceElemType.cast<LLVM::LLVMStructType>(),
+                   convViewElemType, indices, subViewOp, transformed, rewriter);
+    assert(!indices.empty() && "Expecting a least one index");
+
+    // According to MLIRScanner::InitializeValueByInitListExpr() in
+    // clang-mlir.cc, when a memref element type is a struct type, the return
+    // type of a polygeist.subindex operation should be a memref of the
+    // element type of the struct.
+    auto elemPtrTy = LLVM::LLVMPointerType::get(convViewElemType);
+    auto gep = rewriter.create<LLVM::GEPOp>(loc, elemPtrTy, prev, indices);
+    auto memRefDesc = createMemRefDescriptor(loc, viewMemRefType, gep, gep,
+                                             sizes, strides, rewriter);
+    LLVM_DEBUG(llvm::dbgs() << "gep: " << *gep << "\n");
```

Comment:
This is the only debug output in this file, can you please add a prefix to make it clearer?

Reply:
ok